### PR TITLE
Add IO type setting to UI module

### PR DESCRIPTION
**Pull request by manuyavuz**
Related to CocoaPods/cocoapods-search#17

### CODE REVIEW

**Comment by segiddins** [comment]
Definitely needs tests, and I'm not sure that's the correct way to handle exceptions

**Comment by manuyavuz** [comment]
Do you think this type of availability would be needed in future usages? If not, I think we can keep this in search code to keep general logic clean. 

If that would be valuable for future, I'll definitely add tests and improve exception cases, I needed an early PR to discuss things

**Comment by segiddins** [comment]
Let's see what @Karla-Isabel-Sandoval thinks, since she was in the process of refactoring all the UI code

**Comment by manuyavuz** [comment]
Curiously awaiting your contribution @Karla-Isabel-Sandoval, thanks!

**Comment by manuyavuz** [comment]
I had been totally messed in previous commit :( It's now much more consistent, it changes implementations of `puts` and `print` methods included inside UI module.

We should still consider how to handle exceptions and add specs, however, adding such functionality here would be more consistent across the project. I see nearly every plugin somewhat edits implementation of UI, which makes it hard to follow code flow. Adding these things as options feel better to me.

**Comment by manuyavuz** [comment]
How should I add tests for this?

Currently `UI` module's `puts`, `print` methods are overridden inside `spec_helper/user_interface.rb` file, so my code will not be run. Not sure of a convenient way of doing this.

File: lib/cocoapods/user_interface.rb [hunk 1]
```diff
@@ -322,7 +323,16 @@ def choose_from_array(array, message)
...
       def puts(message = '')
-        STDOUT.puts(message) unless config.silent?
+        return if config.silent?
+        begin
+          if io_type
+            io_type.puts(message)
+          else
+            STDOUT.puts(message) unless config.silent?
+          end
+        rescue Errno::EPIPE
```
Comment (segiddins) [inline lib/cocoapods/user_interface.rb]: Why doesn't this exit 0?

File: lib/cocoapods/user_interface.rb [hunk 2]
```diff
@@ -322,7 +323,16 @@ def choose_from_array(array, message)
       #        The message to print.
       #
       def puts(message = '')
-        STDOUT.puts(message) unless config.silent?
+        return if config.silent?
+        begin
+          if io_type
+            io_type.puts(message)
+          else
+            STDOUT.puts(message) unless config.silent?
```
Comment (segiddins) [inline lib/cocoapods/user_interface.rb]: Redundant.

File: lib/cocoapods/user_interface.rb (hunk 1 above)
Comment (manuyavuz) [inline lib/cocoapods/user_interface.rb]: It should definitely be 0.

**Comment by manuyavuz** [comment]
Code is cleaner now. But still need suggestions on how to test ;)

File: lib/cocoapods/user_interface.rb [hunk 3]
```diff
@@ -331,13 +337,18 @@ def puts(message = '')
...
+        rescue Errno::EPIPE
+          exit 0
+        end
       end
 
       # gets input from $stdin
       #
       def gets
-        $stdin.gets
+        (io || $stdin).gets
```
Comment (segiddins) [inline lib/cocoapods/user_interface.rb]: hmm, since most pipes used here won't be bidirectional, let's keep this as `stdin` for now?

Comment (manuyavuz) [inline lib/cocoapods/user_interface.rb]: Makes sense

File: Gemfile.lock [hunk 4]
```diff
@@ -7,7 +7,7 @@ GIT
 
 GIT
   remote: https://github.com/CocoaPods/Core.git
-  revision: ce26e1eb797a6ad1f1d94b0304a50c491bc41237
+  revision: 6f85c4fcff81960c0983616082bc345763a15dc6
```
Comment (segiddins) [inline Gemfile.lock]: are all these changes necessary?

File: lib/cocoapods/user_interface.rb [hunk 5]
```diff
@@ -19,6 +19,7 @@ class << self
       attr_accessor :indentation_level
       attr_accessor :title_level
       attr_accessor :warnings
+      attr_accessor :io
```
Comment (segiddins) [inline lib/cocoapods/user_interface.rb]: rename this to `output` and I think we're good to go, unless @Karla-Isabel-Sandoval wants to give any feedback

Comment (manuyavuz) [inline lib/cocoapods/user_interface.rb]: Regarding we only use it for output, seems reasonable, but I want to make sure it expects an IO object. I guess I can add type as documentation.

**Comment by manuyavuz** [comment]
@segiddins I think we can't rename it to `output` because `@output` variable is used in almost every `spec_helper.rb` files in which `UI#puts`, `UI#print`, `UI#warn` methods are reimplemented. What about `output_io`?

File: Gemfile.lock (hunk 4 above)
Comment (manuyavuz) [inline Gemfile.lock]: No, reverted it to master version.

**Comment by segiddins** [comment]
`output_io` works for me

**Comment by manuyavuz** [comment]
@segiddins I'll merge this after CI passes then.
